qiita_db/metadata_template/base_metadata_template.py

```diff
@@ -1255,47 +1309,26 @@ def _update(self, md_template):
             # between the current_map and md_template
             changed_to = md_template.values[np.where(diff_map)]
 
-            # to_update is a MultiIndexed DataFrame, in which the index 0 is
-            # the samples and the index 1 is the columns, we define these
-            # variables here so we don't put magic numbers across the code
-            sample_idx = 0
-            col_idx = 1
             to_update = pd.DataFrame({'to': changed_to}, index=changed.index)
 
-            # Get the columns that we need to change
-            indices = list(set(to_update.index.labels[col_idx]))
-            cols_to_update = to_update.index.levels[col_idx][indices]
-
-            if not self.can_be_updated(columns=set(cols_to_update)):
-                raise qdb.exceptions.QiitaDBError(
-                    'The new template is modifying fields that cannot be '
-                    'modified. Try removing the restricted fields or '
-                    'deleting the processed data. You are trying to modify: %s'
-                    % ', '.join(cols_to_update))
-
-            # Get the samples that we need to change
-            indices = list(set(to_update.index.labels[sample_idx]))
-            samples_to_update = to_update.index.levels[sample_idx][indices]
-
-            sql_eq_cols = ', '.join(
-                ["{0} = c.{0}".format(col) for col in cols_to_update])
-            # We add 1 because we need to add the sample name
-            single_value = "(%s)" % ', '.join(
-                ["%s"] * (len(cols_to_update) + 1))
-            sql_cols = ', '.join(cols_to_update)
-
-            sql = """UPDATE qiita.{0} AS t SET
-                        {1}
-                     FROM (VALUES {2})
-                        AS c(sample_id, {3})
-                     WHERE c.sample_id = t.sample_id
-                    """.format(self._table_name(self._id), sql_eq_cols,
-                               single_value, sql_cols)
-            for sample in samples_to_update:
-                sample_vals = [md_template[col][sample]
-                               for col in cols_to_update]
-                sample_vals.insert(0, sample)
-                qdb.sql_connection.TRN.add(sql, sample_vals)
+            new_columns = []
+            for sid, df in to_update.groupby(level=0):
+                values = {k[1]: v for k, v in df.to_dict()['to'].iteritems()}
+                new_columns.extend(values.keys())
+                sql = """UPDATE qiita.{0}
+                         SET sample_values = sample_values || %s
+                         WHERE sample_id = %s""".format(
+                            self._table_name(self._id))
+                qdb.sql_connection.TRN.add(sql, [dumps(values), sid])
+
+            new_columns = list(set(new_columns).union(set(self.categories())))
+            table_name = self._table_name(self.id)
+            values = '{"columns": %s}' % dumps(new_columns)
```

Comment:
same same. may be better to decompose this into a helper method since this pattern has come up a bunch

Reply:
👍